File: src/main/java/library/autodispose/observable/ObservableProxy.java
package library.autodispose.observable;

import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;

public interface ObservableProxy<T> {

    @NonNull
    Disposable subscribe(
            @NonNull Consumer<? super T> onNext,
            @NonNull Consumer<? super Throwable> onError,
            @NonNull Action onComplete,
            @NonNull Consumer<? super Disposable> onSubscribe);

    @NonNull
    Disposable subscribe(
            @NonNull Consumer<? super T> onNext,
            @NonNull Consumer<? super Throwable> onError,
            @NonNull Action onComplete);

    @NonNull
    Disposable subscribe(
            @NonNull Consumer<? super T> onNext,
            @NonNull Consumer<? super Throwable> onError);

    @NonNull
    Disposable subscribe(
            @NonNull Consumer<? super T> onNext);
}

File: src/main/java/library/autodispose/observable/ObservableProxy.kt
package library.autodispose.observable

import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer

interface ObservableProxy<T> {

    fun subscribe(
            onNext: Consumer<T>? = null,
            onError: Consumer<Throwable>? = null,
            onComplete: Action? = null,
            onSubscribe: Consumer<Disposable>? = null): Disposable

    fun doOnSubscribe(
            onSubscribe: Consumer<Disposable>): ObservableProxy<T>
}